tests/unit/compiler/venom/test_algebraic_optimizer.py

```diff
@@ -1,182 +1,218 @@
+def test_simple_bool_cast_case(iszero_count):
+    """
+    Test that iszero chain elimination would not eliminate
+    bool cast
 
-    assert removed_iszeros % 2 == 0
-    assert len(iszeros) == iszero_count % 2
+    sink(izero(iszero(iszero(iszero(x))))) => sink(iszero(iszero(x)))
 
+    You cannot remove all iszeros because the sink expects the bool
+    and the total elimination would invalidate it
+    """
 
-@pytest.mark.parametrize("iszero_count", range(1, 5))
-def test_simple_bool_cast_case(iszero_count):
-    ctx = IRContext()
-    fn = ctx.create_function("_global")
+    iszero_chain = ""
+    for i in range(iszero_count):
+        new = i + 1
+        iszero_chain += f"""
+        %cond{new} = iszero %cond{i}"""
+
+    pre = f"""
+    main:
+        %par = param
+        %1 = %par
+        %2 = 64
+        %3 = add %1, %2
+        %cond0 = %3
+        {iszero_chain}
+        sink %cond{iszero_count}
+    """
 
-    bb = fn.get_basic_block()
+    if iszero_count % 2 == 0:
+        post_chain = """
+        %cond1 = iszero %cond0
+        %cond2 = iszero %cond1
+        """
+        end_cond = "cond2"
+    else:
+        post_chain = """
+        %cond1 = iszero %cond0
+        """
+        end_cond = "cond1"
+
+    post = f"""
+    main:
+        %par = param
+        %1 = %par
+        %2 = 64
+        %3 = add %1, %2
+        %cond0 = %3
+        {post_chain}
+        sink %{end_cond}
+    """
 
-    br1 = IRBasicBlock(IRLabel("then"), fn)
-    fn.append_basic_block(br1)
+    _check_pre_post(pre, post)
 
-    p1 = bb.append_instruction("param")
-    op1 = bb.append_instruction("store", p1)
-    op2 = bb.append_instruction("store", 64)
-    op3 = bb.append_instruction("add", op1, op2)
-    jnz_input = op3
 
-    for _ in range(iszero_count):
-        jnz_input = bb.append_instruction("iszero", jnz_input)
+@pytest.mark.parametrize("interleave_point", range(5))
+def test_interleaved_case(interleave_point):
+    """
+    Test for the case where one of the iszeros results in
+    the chain is used
+    """
 
-    bb.append_instruction("mstore", jnz_input, p1)
-    bb.append_instruction("jmp", br1.label)
+    iszeros_after_interleave_point = interleave_point // 2
 
-    br1.append_instruction("add", op3, 10)
-    br1.append_instruction("stop")
+    before_iszeros = ""
+    for i in range(interleave_point):
+        new = i + 2
+        before_iszeros += f"""
+        %cond{new} = iszero %cond{i + 1}"""
+
+    after_iszeros = ""
+    for i in range(iszeros_after_interleave_point):
+        index = i + interleave_point + 1
+        new = index + 1
+        after_iszeros += f"""
+        %cond{new} = iszero %cond{index}"""
+
+    pre = f"""
+    main:
+        %par = param
+        %cond0 = add 64, %par
+        %cond1 = iszero %cond0
+        {before_iszeros}
+        mstore %par, %cond{interleave_point + 1}
+        {after_iszeros}
+        jnz %cond{interleave_point + iszeros_after_interleave_point + 1}, @then, @else
```

Comment:
maybe move this calculation to outside of the string construction, and call it `{jnz_cond}`